### DISCUSSION TITLE
The ouput of the yolov8 model is throwing error to display

### DISCUSSION THREAD

**Discussion by subhakarkoduru** [ultralytics/ultralytics#8211] [Q&A] [answered]
Hi, I had trained a yolov8 model and trying to predict the img using that model, but getting error when i am trying to display it 

results = model(['img1.jpg'])  # return a list of Results objects

# Process results list
for result in results:
    boxes = result.boxes  # Boxes object for bounding box outputs
    masks = result.masks  # Masks object for segmentation masks outputs
    keypoints = result.keypoints  # Keypoints object for pose outputs
    probs = result.probs  # Probs object for classification outputs
    result.show()  # display to screen
    result.save(filename='result.jpg')  # save to disk
    
    
    error :

AttributeError: 'Results' object has no attribute 'show'. See valid attributes below.

    A class for storing and manipulating inference results.

    Args:
        orig_img (numpy.ndarray): The original image as a numpy array.
        path (str): The path to the image file.
        names (dict): A dictionary of class names.
        boxes (torch.tensor, optional): A 2D tensor of bounding box coordinates for each detection.
        masks (torch.tensor, optional): A 3D tensor of detection masks, where each mask is a binary image.
        probs (torch.tensor, optional): A 1D tensor of probabilities of each class for classification task.
        keypoints (List[List[float]], optional): A list of detected keypoints for each object.

    Attributes:
        orig_img (numpy.ndarray): The original image as a numpy array.
        orig_shape (tuple): The original image shape in (height, width) format.
        boxes (Boxes, optional): A Boxes object containing the detection bounding boxes.
        masks (Masks, optional): A Masks object containing the detection masks.
        probs (Probs, optional): A Probs object containing probabilities of each class for classification task.
        keypoints (Keypoints, optional): A Keypoints object containing detected keypoints for each object.
        speed (dict): A dictionary of preprocess, inference, and postprocess speeds in milliseconds per image.
        names (dict): A dictionary of class names.
        path (str): The path to the image file.
        _keys (tuple): A tuple of attribute names for non-empty attributes.

**Comment by glenn-jocher** [comment, marked as answer]
@subhakarkoduru hello! It looks like you're trying to use a `show()` method on the `Results` object, which doesn't exist. To display the image, you can use the `plot()` method to visualize the predictions and then use OpenCV or PIL to show it. Here's how you can modify your code:

```python
import cv2  # Import OpenCV

# ... [your existing code] ...

for result in results:
    # ... [your existing code to get boxes, masks, etc.] ...
    img = result.plot()  # Plot the predictions on the image
    cv2.imshow('Predictions', img)  # Display the image using OpenCV
    cv2.waitKey(0)  # Wait for a key press to close the window
    cv2.destroyAllWindows()  # Close the window
    cv2.imwrite('result.jpg', img)  # Save the image to disk
```

Make sure you have OpenCV installed (`pip install opencv-python`) to use the `cv2` module. If you prefer using PIL to display the image, you can convert the numpy array to a PIL Image and then show it. Hope this helps! 😊👍

**Comment by subhakarkoduru** [comment]
What the Confidence Therhold used by yolov8 model to determain the object is dected or not ?

**Reply by glenn-jocher** [reply]
Hello! The confidence threshold in YOLOv8 determines the minimum probability required for a detection to be considered valid. By default, during prediction, a confidence threshold of 0.25 is used. However, you can adjust this threshold based on your specific needs. For example, to use a confidence threshold of 0.5, you can set the `conf` argument like this:

```python
results = model.predict(source='image.jpg', conf=0.5)
```

This will filter out detections with confidence scores below 0.5. Adjust the threshold to balance between precision and recall for your use case. 😊

**Comment by subhakarkoduru** [comment]
Thanks for your Inputs, I have a small question In a certain senario i am getting two bounding boxes for same object what can i do so solve this issue.

<img width="1470" alt="Screenshot 2024-02-16 at 11 50 23 AM" src="https://github.com/ultralytics/ultralytics/assets/31570443/415d522b-7529-4123-a4a7-40fc582e6566">

**Reply by pderrenger** [reply]
Hello! It looks like you're experiencing overlapping bounding boxes for the same object, which can happen due to multiple detections. To resolve this, you can adjust the Non-Maximum Suppression (NMS) threshold. A higher NMS threshold (`iou`) may help reduce these duplicate detections by being more aggressive in filtering out overlapping boxes.

Here's a quick example of how you can adjust the NMS threshold during prediction:

```python
results = model(['img1.jpg'], iou=0.7)  # Increase the IoU threshold for NMS
```

Give this a try and see if it helps with your scenario! If you're still facing issues, feel free to share more details, and we'll be happy to assist further. 😊👍

**Reply by subhakarkoduru** [reply]
Hi pderrenger,

 I think the lowering the Iou is working iou=0.6 is there any reason for that. Can i add iou and conf parameters during training process of the model

**Reply by pderrenger** [reply]
@subhakarkoduru hi there,

Glad to hear that adjusting the IoU threshold is helping! Lowering the IoU threshold can reduce duplicate detections because it increases the overlap required for boxes to be considered the same object.

Yes, you can indeed set `iou` and `conf` thresholds during training for the NMS process. Here's an example of how you might adjust these parameters in your training command:

```bash
yolo train data=your_dataset.yaml model=yolov8n.yaml iou_t=0.6 conf_t=0.25
```

Just replace `your_dataset.yaml` and `yolov8n.yaml` with your actual dataset and model configuration files. Keep in mind that these thresholds are used during the evaluation phase in training to calculate metrics like mAP.

Happy training! 😊🚀
